server-timing/defaults.php

```diff
@@ -0,0 +1,178 @@
+				$calculate_before_template_metrics();
+			}
+			return $passthrough;
+		},
+		PHP_INT_MAX
+	);
+	add_action(
+		'perflab_server_timing_send_header',
+		function() use ( $calculate_before_template_metrics ) {
+			if ( ! perflab_server_timing_use_output_buffer() ) {
+				$calculate_before_template_metrics();
+			}
+		},
+		PHP_INT_MAX
+	);
+
+	// Measure duration of autoloaded options query.
+	// Requires the Performance Lab object-cache.php drop-in to be present in order to work.
+	add_filter(
+		'query',
+		function( $query ) {
+			global $wpdb;
+			if ( "SELECT option_name, option_value FROM $wpdb->options WHERE autoload = 'yes'" !== $query ) {
+				return $query;
+			}
+			perflab_server_timing_register_metric(
+				'load-alloptions-query',
+				array(
+					'measure_callback' => function( $metric ) {
+						$metric->measure_before();
+						add_filter(
+							'pre_cache_alloptions',
+							function( $passthrough ) use ( $metric ) {
+								$metric->measure_after();
+								return $passthrough;
+							}
+						);
+					},
+					'access_cap'       => 'exist',
+				)
+			);
+			return $query;
+		}
+	);
+}
+perflab_register_default_server_timing_before_template_metrics();
+
+/**
+ * Registers the default Server-Timing metrics for while rendering the template.
+ *
+ * These metrics should be registered at a later point, e.g. the 'wp_loaded' action.
+ * They will only be registered if the Server-Timing API is configured to use an
+ * output buffer for the site's template.
+ *
+ * @since n.e.x.t
+ */
+function perflab_register_default_server_timing_template_metrics() {
+	// Template-related metrics can only be recorded if output buffering is used.
+	if ( ! perflab_server_timing_use_output_buffer() ) {
+		return;
+	}
+
+	add_filter(
+		'template_include',
+		function( $passthrough = null ) {
+			// WordPress execution while serving the template.
+			perflab_server_timing_register_metric(
+				'template',
+				array(
+					'measure_callback' => function( $metric ) {
+						$metric->measure_before();
+						add_action( 'perflab_server_timing_send_header', array( $metric, 'measure_after' ), PHP_INT_MAX );
+					},
+					'access_cap'       => 'exist',
+				)
+			);
+
+			return $passthrough;
+		},
+		PHP_INT_MAX
+	);
+
+	if ( defined( 'SAVEQUERIES' ) && SAVEQUERIES ) {
+		add_action(
+			'perflab_server_timing_send_header',
+			function() {
+				// WordPress database query time within template.
+				perflab_server_timing_register_metric(
+					'template-db-queries',
+					array(
+						'measure_callback' => function( $metric ) {
+							// This global should typically be set when this is called, but check just in case.
+							if ( ! isset( $GLOBALS['perflab_query_time_before_template'] ) ) {
+								return;
+							}
+
+							$total_query_time = array_reduce(
+								$GLOBALS['wpdb']->queries,
+								function( $acc, $query ) {
+									return $acc + $query[1];
+								},
+								0.0
+							);
```

Comment:
@felixarntz same problem here, we need to check whether the `$GLOBALS['wpdb']->queries` is not NULL.

Reply:
@eugene-manuilov Technically I could add the extra check, but the one above should already prevent that problem (since that global only gets set if `$GLOBALS['wpdb']->queries` is an array in the metric measure callback further up, which runs before.

Reply:
@eugene-manuilov Good idea, will add this.